06-job-control.md

```diff
@@ -4,8 +4,169 @@ title: Extra Unix Shell Material
+
+~~~{.input}
+$ ./analyze results*.dat
+~~~
+
+~~~
+...a few minutes pass...
+^C
+~~~
+
+Let's run that same command again, with an ampersand `&` at the end of
+the line to tell the shell we want it to run in the
+[background](glossary.html#background):
+
+~~~{.input}
+$ ./analyze results*.dat &
+~~~
+
+When we do this, the shell launches the program as before. Instead of
+leaving our keyboard and screen connected to the program's standard
+input and output, though, the shell hangs onto them. This means the
+shell can give us a fresh command prompt, and start running other
+commands, right away. Here, for example, we're putting some parameters
+for the next run of the program in a file:
+
+~~~{.input}
+$ cat > params.txt
+~~~
+~~~{.output}
+density: 22.0
+viscosity: 0.75
+^D
+~~~
+
+(Remember, \^D is the shell's way of showing Control-D, which means "end
+of input".) Now let's run the `jobs` command, which tells us what
+processes are currently running in the background:
+
+~~~{.input}
+$ jobs
+~~~
+~~~{.output}
+[1] ./analyze results01.dat results02.dat results03.dat
+~~~
+
+Since we're about to go and get coffee, we might as well use the
+foreground command, `fg`, to bring our background job into the
+foreground:
+
+~~~{.input}
+$ fg
+~~~
+~~~
+...a few minutes pass...
+~~~
+
+When `analyze` finishes running, the shell gives us a fresh prompt as
+usual. If we had several jobs running in the background, we could
+control which one we brought to the foreground using `fg %1`, `fg %2`,
+and so on. The IDs are *not* the process IDs. Instead, they are the job
+IDs displayed by the `jobs` command.
+
+The shell gives us one more tool for job control: if a process is
+already running in the foreground, Control-Z will pause it and return
+control to the shell. We can then use `fg` to resume it in the
+foreground, or `bg` to resume it as a background job. For example, let's
+run `analyze` again, and then type Control-Z. The shell immediately
+tells us that our program has been stopped, and gives us its job number:
+
+~~~{.input}
+$ ./analyze results01.dat
+
+^Z
+~~~
+~~~{.output}
+[1]  Stopped   ./analyze results01.dat
+~~~
+
+If we type `bg %1`, the shell starts the process running again, but in
+the background. We can check that it's running using `jobs`, and kill it
+while it's still in the background using `kill` and the job number. This
+has the same effect as bringing it to the foreground and then typing
+Control-C:
+
+~~~{.input}
+$ bg %1
+$ jobs
+~~~
+~~~{.output}
+[1] ./analyze results01.dat
+~~~
+~~~{.input}
+$ kill %1
+~~~
+
+Job control was important when users only had one terminal window at a
+time. It's less important now: if we want to run another program, it's
+easy enough to open another window and run it there. However, these
+ideas and tools are making a comeback, as they're often the easiest way
+to run and control programs on remote computers elsewhere on the
+network. 
```

Comment:
The "easiest way" is rather hand wavy. These tools help if you only want to have one SSH terminal session open, is that what is meant here?